### PR TITLE
Update Intl object for support new Intl spec

### PR DESCRIPTION
**Pull request by ksh8281**

### CODE REVIEW

File: src/intl/Intl.cpp [hunk 1]
```diff
@@ -1808,6 +1896,50 @@ Value Intl::getOption(ExecutionState& state, Object* options, Value property, In
...
+    // If value is not undefined, then
+    if (!value.isUndefined()) {
+        // Let value be ToNumber(value).
+        double doubleValue = value.toNumber(state);
+        // If value is NaN or less than minimum or greater than maximum, throw a RangeError exception.
+        if (std::isnan(doubleValue) || doubleValue < minimum || maximum < doubleValue) {
+            ErrorObject::throwBuiltinError(state, ErrorObject::RangeError, "Got invalid number option value");
+        }
+        // Return floor(value).
+        return Value(floor(doubleValue));
```
Comment (bbrto21) [inline src/intl/Intl.cpp:1936]: The only difference between the two implementations of `Intl::getNumberOption` is this.
I think you can extract a common part as a new method

File: src/intl/IntlDateTimeFormat.cpp [hunk 2]
```diff
@@ -275,34 +268,77 @@ UTF16StringDataNonGCStd updateHourCycleInPatternDueToHourCycle(const UTF16String
...
 
 
-void IntlDateTimeFormat::initialize(ExecutionState& state, Object* dateTimeFormat, Value locales, Value options)
+static String* icuFieldTypeToPartName(ExecutionState& state, int32_t fieldName)
 {
-    // If dateTimeFormat has an [[initializedIntlObject]] internal property with value true, throw a TypeError exception.
-    AtomicString initializedIntlObject = state.context()->staticStrings().lazyInitializedIntlObject();
-    if (dateTimeFormat->hasInternalSlot() && dateTimeFormat->internalSlot()->hasOwnProperty(state, ObjectPropertyName(state, ObjectStructurePropertyName(initializedIntlObject)))) {
-        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "Cannot initialize Intl Object twice");
+    if (fieldName == -1) {
```
Comment (bbrto21) [inline src/intl/IntlDateTimeFormat.cpp:273]: What does `-1` mean? and why not use `< 0`?

File: src/intl/IntlDateTimeFormat.cpp [hunk 3]
```diff
@@ -275,34 +268,77 @@ UTF16StringDataNonGCStd updateHourCycleInPatternDueToHourCycle(const UTF16String
...
+        return state.context()->staticStrings().lazyLiteral().string();
+    }
+
+    switch ((UDateFormatField)fieldName) {
```
Comment (bbrto21) [inline src/intl/IntlDateTimeFormat.cpp:277]: Please avoid pure c-style casting

File: src/intl/IntlDateTimeFormat.cpp [hunk 4]
```diff
@@ -544,17 +601,174 @@ void IntlDateTimeFormat::initialize(ExecutionState& state, Object* dateTimeForma
...
+
+    } else {
+        // Else,
+        // Let formats be dataLocaleData.[[formats]].[[<calendar>]].
+        // If matcher is "basic", then
+        // Let bestFormat be BasicFormatMatcher(opt, formats).
+        // Else,
+        // Let bestFormat be BestFitFormatMatcher(opt, formats).
+        skeleton = skeletonBuilder.finalize()->toUTF16StringData();
+        patternBuffer.resize(32);
```
Comment (bbrto21) [inline src/intl/IntlDateTimeFormat.cpp:758]: `32` is used multiple times. How about giving it a name?

File: src/builtins/BuiltinIntl.cpp [hunk 5]
```diff
@@ -972,7 +982,7 @@ void GlobalObject::installIntl(ExecutionState& state)
     m_intlDateTimeFormatPrototype = m_intlDateTimeFormat->getFunctionPrototype(state).asObject();
     m_intlDateTimeFormatPrototype->setGlobalIntrinsicObject(state, true);
 
-    FunctionObject* formatFunction = new NativeFunctionObject(state, NativeFunctionInfo(strings->getFormat, builtinIntlDateTimeFormatFormatGetter, 0, NativeFunctionInfo::Strict));
+    FunctionObject* formatFunction = new NativeFunctionObject(state, NativeFunctionInfo(state.context()->staticStrings().getFormat, builtinIntlDateTimeFormatFormatGetter, 0, NativeFunctionInfo::Strict));
```
Comment (bbrto21) [inline src/builtins/BuiltinIntl.cpp:985]: Why do you get a staticStrings reference again?

File: src/builtins/BuiltinIntl.cpp [hunk 6]
```diff
@@ -990,9 +1000,12 @@ void GlobalObject::installIntl(ExecutionState& state)
...
     m_intlNumberFormatPrototype->setGlobalIntrinsicObject(state, true);
 
-    formatFunction = new NativeFunctionObject(state, NativeFunctionInfo(strings->getFormat, builtinIntlNumberFormatFormatGetter, 0, NativeFunctionInfo::Strict));
-    m_intlNumberFormatPrototype->defineOwnProperty(state, state.context()->staticStrings().format,
-                                                   ObjectPropertyDescriptor(JSGetterSetter(formatFunction, Value(Value::EmptyValue)), (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::ConfigurablePresent)));
+    {
+        formatFunction = new NativeFunctionObject(state, NativeFunctionInfo(strings->getFormat, builtinIntlNumberFormatFormatGetter, 0, NativeFunctionInfo::Strict));
+        JSGetterSetter gs(formatFunction, Value());
+        ObjectPropertyDescriptor desc(gs, ObjectPropertyDescriptor::ConfigurablePresent);
+        m_intlNumberFormatPrototype->defineOwnProperty(state, ObjectPropertyName(state, state.context()->staticStrings().format), desc);
```
Comment (bbrto21) [inline src/builtins/BuiltinIntl.cpp:1007]: Same here.

File: src/intl/IntlDateTimeFormat.cpp (hunk 2 above)
Comment (ksh8281) [reply src/intl/IntlDateTimeFormat.cpp:273]: It's special value for ICU

File: src/intl/IntlDateTimeFormat.cpp (hunk 3 above)
Comment (ksh8281) [reply src/intl/IntlDateTimeFormat.cpp:277]: Good advice

File: src/intl/Intl.cpp (hunk 1 above)
Comment (ksh8281) [reply src/intl/Intl.cpp:1936]: I made a special method for it

File: src/builtins/BuiltinIntl.cpp (hunk 6 above)
Comment (ksh8281) [reply src/builtins/BuiltinIntl.cpp:1007]: Good advice

File: src/builtins/BuiltinIntl.cpp (hunk 5 above)
Comment (ksh8281) [reply src/builtins/BuiltinIntl.cpp:985]: Good advice

**Review by bbrto21 — APPROVED** [review]

**Review by clover2123 — APPROVED** [review]
LGTM

**Pull request merged**